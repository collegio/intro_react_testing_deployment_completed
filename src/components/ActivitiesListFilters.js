import React from 'react';
import { connect } from 'react-redux';
import { setFilterText, setFilterType, sortByName, sortByDistance } from '../actions/filters';

export class ActivitiesListFilters extends React.Component {

    constructor(props) {
        super(props);

        this.onTextChange = this.onTextChange.bind(this);
        this.onSortChange = this.onSortChange.bind(this);
        this.onActivityTypeChange = this.onActivityTypeChange.bind(this);
    }

    onTextChange(e) {
        this.props.setFilterText(e.target.value);
    }

    onSortChange(e) {
        if (e.target.value === "name") {
            this.props.sortByName();
        }
        else if (e.target.value === "distance") {
            this.props.sortByDistance();
        }
    }

    onActivityTypeChange(e) {
        this.props.setFilterType(e.target.value);
    }

    render() {
        return (
            <div className="activities-list-filters">
                <h2>Filter Activities</h2>
                <div className="filter-group">
                    <label>Search For:</label><br />
                    <input type="text" value={this.props.filters.text} onChange={this.onTextChange} />
                </div>
                <div className="filter-group">
                    <label>Sort By:</label><br />
                    <select onChange={this.onSortChange}>
                        <option value="name">Name</option>
                        <option value="distance">Distance</option>
                    </select>
                </div>
                <div className="filter-group">
                    <label>Sport Type:</label><br />
                    <select onChange={this.onActivityTypeChange}>
                        <option value="all">All</option>
                        <option value="running">Running</option>
                        <option value="walking">Walking</option>
                        <option value="cycling">Cycling</option>
                        <option value="swimming">Swimming</option>
                    </select>
                </div>
            </div>
        );
    }
}

const mapStateToProps = (state) => {
    return {
        filters: state.filters
    };
};

const mapDispatchToProps = (dispatch) => {
    return {
        setFilterText: (text) => dispatch(setFilterText(text)),
        setFilterType: (activity_type) => dispatch(setFilterType(activity_type)),
        sortByName: () => dispatch(sortByName()),
        sortByDistance: () => dispatch(sortByDistance()),
    };
};

export default connect(mapStateToProps, mapDispatchToProps)(ActivitiesListFilters);