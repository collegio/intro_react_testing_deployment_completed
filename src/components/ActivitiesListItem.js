import React from 'react';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import { removeActivity } from '../actions/activities';

export const ActivitiesListItem = ({ dispatch, id, name, activity_type, distance }) => (
    <div className="activity">
        <div className="activity-info">
            <p>
                <Link to={'/edit/'+id}>
                    {name}<br />
                </Link>
                {activity_type} - {distance}
            </p>
            <button className="button button-clear" onClick={() => {
                dispatch(removeActivity({ id }));
            }}>Remove</button>
        </div>
    </div>
);

export default connect()(ActivitiesListItem);