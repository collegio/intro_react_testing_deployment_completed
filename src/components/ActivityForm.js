import React from 'react';

export class ActivityForm extends React.Component {

    constructor(props) {
        super(props);

        this.onNameChange = this.onNameChange.bind(this);
        this.onActivityChange = this.onActivityChange.bind(this);
        this.onDistanceChange = this.onDistanceChange.bind(this);
        this.onSubmitActivity = this.onSubmitActivity.bind(this);

        this.state = {
            name: props.activity ? props.activity.name : '',
            activity_type: props.activity ? props.activity.activity_type : 'walking',
            distance: props.activity ? props.activity.distance : '0 km',
            error: ''
        }
    }

    onNameChange(e) {
        const name = e.target.value;
        this.setState(() => ({ name }));
    };

    onActivityChange(e) {
        const activity_type = e.target.value;
        this.setState(() => ({ activity_type }));
    };

    onDistanceChange(e) {
        const distance = e.target.value;
        this.setState(() => ({ distance }));
    };

    onSubmitActivity(e) {
        e.preventDefault();

        if (!this.state.name) {

            // Error!
            this.setState(() => ({ error: 'Please enter a name first.' }));
        }
        else {

            // Remove the error and Submit!
            this.setState(() => ({ error: '' }));
            this.props.onSubmit({
                name: this.state.name,
                activity_type: this.state.activity_type,
                distance: this.state.distance
            });
        }
    }

    render() {
        return (
            <div className="activity-form">
                { this.state.error && <p>{this.state.error}</p> }
                <form onSubmit={this.onSubmitActivity}>
                    <label>Name</label><br/>
                    <input type="text" placeholder="Name" value={this.state.name} onChange={this.onNameChange} autoFocus />

                    <label>Sport</label><br/>
                    <select value={this.state.sport_type} onChange={this.onActivityChange}>
                        <option value="running">Running</option>
                        <option value="walking">Walking</option>
                        <option value="cycling">Cycling</option>
                        <option value="swimming">Swimming</option>
                    </select>

                    <label>Distance</label><br/>
                    <input type="text" placeholder="Name" value={this.state.distance} onChange={this.onDistanceChange} autoFocus />

                    <button className="button pull-right">Submit</button>
                </form>
            </div>
        );
    }
}

export default ActivityForm;