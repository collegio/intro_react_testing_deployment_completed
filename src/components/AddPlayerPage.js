import React from 'react';
import { connect } from 'react-redux';
import PlayerForm from './PlayerForm';
import { addPlayer } from '../actions/players';

export class AddPlayerPage extends React.Component {

    constructor(props) {
        super(props);

        this.onSubmit = this.onSubmit.bind(this);
    }

    onSubmit(player) {
        this.props.addPlayer(player);
        this.props.history.push('/');
    }

    render() {
        return (
            <div className="container">
                <h1>Add a Player</h1>
                <PlayerForm 
                    onSubmit={this.onSubmit}
                />
            </div>
        );
    }
}

const mapDispatchToProps = (dispatch) => ({
    addPlayer: (player) => dispatch(addPlayer(player))
});

export default connect(undefined, mapDispatchToProps)(AddPlayerPage);