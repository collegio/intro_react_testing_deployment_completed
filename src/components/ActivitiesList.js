import React from 'react';
import { connect } from 'react-redux';
import ActivitiesListItem from './ActivitiesListItem';
import selectActivities from '../selectors/activities';

export const ActivitiesList = (props) => (
    <div className="activities-list">
        <h1>Activities List</h1>
        <div className="activity-list">
            {props.activities.map((activity) => {
                return <ActivitiesListItem key={activity.id} {...activity} />;
            })}
        </div>
    </div>
);

const mapStateToProps = (state) => {
    return {
        activities: selectActivities(state.activities, state.filters)
    }
};

export default connect(mapStateToProps)(ActivitiesList);