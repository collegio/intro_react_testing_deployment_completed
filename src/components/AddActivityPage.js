import React from 'react';
import { connect } from 'react-redux';
import ActivityForm from './ActivityForm';
import { addActivity } from '../actions/activities';

export class AddActivityPage extends React.Component {

    constructor(props) {
        super(props);

        this.onSubmit = this.onSubmit.bind(this);
    }

    onSubmit(activity) {
        this.props.addActivity(activity);
        this.props.history.push('/');
    }

    render() {
        return (
            <div className="container">
                <h1>Add a Activity</h1>
                <ActivityForm 
                    onSubmit={this.onSubmit}
                />
            </div>
        );
    }
}

const mapDispatchToProps = (dispatch) => ({
    addActivity: (activity) => dispatch(addActivity(activity))
});

export default connect(undefined, mapDispatchToProps)(AddActivityPage);