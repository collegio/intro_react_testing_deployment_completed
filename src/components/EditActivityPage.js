import React from 'react';
import { connect } from 'react-redux';
import ActivityForm from './ActivityForm';
import { editActivity, removeActivity } from '../actions/activities';

export class EditActivityPage extends React.Component {

    constructor(props) {
        super(props);

        this.onSubmit = this.onSubmit.bind(this);
        this.onRemove = this.onRemove.bind(this);
    }

    onSubmit(activity) {
        this.props.editActivity(this.props.activity.id, activity);
        this.props.history.push('/');
    }

    onRemove() {
        this.props.removeActivity({ id: this.props.activity.id });
        this.props.history.push('/');
    }

    render() {
        return (
            <div className="container">
                <h1>Edit Activity</h1>
                <ActivityForm
                    activity={this.props.activity}
                    onSubmit={this.onSubmit}
                />
                <button className="button button-delete" onClick={this.onRemove}>Remove</button>
            </div>
        );
    }
}

const mapStateToProps = (state, props) => {
    return {
        activity: state.activities.find((activity) => activity.id === props.match.params.id)
    };
};

const mapDispatchToProps = (dispatch, props) => ({
    editActivity: (id, activity) => dispatch(editActivity(id, activity)),
    removeActivity: (data) => dispatch(removeActivity(data))
});

export default connect(mapStateToProps, mapDispatchToProps)(EditActivityPage);