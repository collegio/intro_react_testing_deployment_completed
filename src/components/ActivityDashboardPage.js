import React from 'react';
import ActivitiesList from './ActivitiesList';
import ActivitiesListFilters from './ActivitiesListFilters';

export const ActivityDashboardPage = (props) => (
    <div className="container">
        <ActivitiesListFilters />
        <ActivitiesList />
    </div>
);

export default ActivityDashboardPage;