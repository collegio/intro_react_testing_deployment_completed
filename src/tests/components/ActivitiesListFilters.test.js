import React from 'react';
import { shallow } from 'enzyme';
import { ActivitiesListFilters } from '../../components/ActivitiesListFilters';
import { filters, altFilters1, altFilters2 } from '../fixtures/activityFilters';

let setFilterText, setFilterType, sortByName, sortByDistance, wrapper;

beforeEach(() => {
    setFilterText = jest.fn();
    setFilterType = jest.fn();
    sortByName = jest.fn();
    sortByDistance = jest.fn();
    wrapper = shallow(
        <ActivitiesListFilters 
            filters={filters} 
            setFilterText={setFilterText}
            setFilterType={setFilterType}
            sortByName={sortByName}
            sortByDistance={sortByDistance}
        />
    );
})

test('should render the activities list filters with default values', () => {
    expect(wrapper).toMatchSnapshot();
});

test('should render the activities list filters with alternate filters', () => {
    wrapper.setProps({ filters: altFilters1 });
    expect(wrapper).toMatchSnapshot();
});

test('should handle filter text changes', () => {
    const value = "Crosby";
    wrapper.find('input').simulate('change', {
        target: { value }
    });
    expect(setFilterText).toHaveBeenLastCalledWith(value);
});

test('should handle activity type select changes', () => {
    const value = "cycling";
    wrapper.find('select').at(1).simulate('change', {
        target: { value }
    });
    expect(setFilterType).toHaveBeenLastCalledWith(value);
});

test('should set sorting by distance', () => {
    const value = 'distance';
    wrapper.find('select').at(0).simulate('change', {
        target: {value}
    });
    expect(sortByDistance).toHaveBeenCalled();
});

test('should set sorting by name', () => {
    const value = 'name';
    wrapper.setProps({ filters: altFilters1 });
    wrapper.find('select').at(0).simulate('change', {
        target: {value}
    });
    expect(sortByName).toHaveBeenCalled();
});
