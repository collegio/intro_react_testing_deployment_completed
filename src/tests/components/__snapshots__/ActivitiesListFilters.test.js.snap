// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render the activities list filters with alternate filters 1`] = `
<div
  className="activities-list-filters"
>
  <h2>
    Filter Activities
  </h2>
  <div
    className="filter-group"
  >
    <label>
      Search For:
    </label>
    <br />
    <input
      onChange={[Function]}
      type="text"
      value="walk"
    />
  </div>
  <div
    className="filter-group"
  >
    <label>
      Sort By:
    </label>
    <br />
    <select
      onChange={[Function]}
    >
      <option
        value="name"
      >
        Name
      </option>
      <option
        value="distance"
      >
        Distance
      </option>
    </select>
  </div>
  <div
    className="filter-group"
  >
    <label>
      Sport Type:
    </label>
    <br />
    <select
      onChange={[Function]}
    >
      <option
        value="all"
      >
        All
      </option>
      <option
        value="running"
      >
        Running
      </option>
      <option
        value="walking"
      >
        Walking
      </option>
      <option
        value="cycling"
      >
        Cycling
      </option>
      <option
        value="swimming"
      >
        Swimming
      </option>
    </select>
  </div>
</div>
`;

exports[`should render the activities list filters with default values 1`] = `
<div
  className="activities-list-filters"
>
  <h2>
    Filter Activities
  </h2>
  <div
    className="filter-group"
  >
    <label>
      Search For:
    </label>
    <br />
    <input
      onChange={[Function]}
      type="text"
      value=""
    />
  </div>
  <div
    className="filter-group"
  >
    <label>
      Sort By:
    </label>
    <br />
    <select
      onChange={[Function]}
    >
      <option
        value="name"
      >
        Name
      </option>
      <option
        value="distance"
      >
        Distance
      </option>
    </select>
  </div>
  <div
    className="filter-group"
  >
    <label>
      Sport Type:
    </label>
    <br />
    <select
      onChange={[Function]}
    >
      <option
        value="all"
      >
        All
      </option>
      <option
        value="running"
      >
        Running
      </option>
      <option
        value="walking"
      >
        Walking
      </option>
      <option
        value="cycling"
      >
        Cycling
      </option>
      <option
        value="swimming"
      >
        Swimming
      </option>
    </select>
  </div>
</div>
`;
